src/docquery/ext/model.py

```diff
@@ -128,19 +143,55 @@ def forward(
             start_positions = start_positions.clamp(0, ignored_index)
             end_positions = end_positions.clamp(0, ignored_index)
 
-            loss_fct = CrossEntropyLoss(ignore_index=ignored_index)
+            loss_fct = nn.CrossEntropyLoss(ignore_index=ignored_index)
             start_loss = loss_fct(start_logits, start_positions)
             end_loss = loss_fct(end_logits, end_positions)
             total_loss = (start_loss + end_loss) / 2
 
+        token_logits = None
+        if self.config.token_classification:
+            token_logits = self.token_classifier_head(sequence_output)
+
+            if token_labels is not None:
+                # Loss fn expects logits to be of shape (batch_size, num_labels, 512), but model
+                # outputs (batch_size, 512, num_labels), so we need to move the dimensions around
+                # Ref: https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html
+                token_logits_reshaped = torch.movedim(token_logits, source=token_logits.ndim - 1, destination=1)
+                token_loss = nn.CrossEntropyLoss(reduction=self.config.token_classifier_reduction)(
+                    token_logits_reshaped, token_labels
+                )
+
+                total_loss += self.config.token_classifier_constant * token_loss
+
         if not return_dict:
-            output = (start_logits, end_logits) + outputs[2:]
-            return ((total_loss,) + output) if total_loss is not None else output
+            output = (start_logits, end_logits)
+            if self.token_classification:
+                output = output + (token_logits,)
+
+            output = output + outputs[2:]
+
+            if total_loss is not None:
+                output = (total_loss,) + output
+
+            return output
 
         return QuestionAnsweringModelOutput(
             loss=total_loss,
             start_logits=start_logits,
             end_logits=end_logits,
+            token_logits=token_logits,
             hidden_states=outputs.hidden_states,
             attentions=outputs.attentions,
         )
+
+
+# This is a thin wrapper around LayoutLMTokenClassifierForQuestionAnswering that simply instantiates
+# a default value for token_classification. Once we update transformers to be a version that
+# includes the upstremed LayoutLMForQuestionAnswering class, we can remove this.
+class LayoutLMForQuestionAnswering(LayoutLMTokenClassifierForQuestionAnswering):
+    config_class = LayoutLMConfig
+
+    def __init__(self, config, **kwargs):
+        config = copy.deepcopy(config)
+        config.token_classification = False
```

Comment:
Is this not covered by setting the default value for `token_classification` within `LayoutLMTokenClassifierConfig`?

Reply:
It is, this was just a way of being explicit (e.g. if we set it to true by default in the config class, which kind of makes sense given its purpose).